pydicom/framereader.py

```diff
@@ -0,0 +1,1182 @@
+        # end of the compressed byte stream.
+        if first_two_bytes in _START_MARKERS:
+            current_offset = frame_position - first_frame_location
+            offset_values.append(current_offset)
+
+        i += 1
+        fp.seek(length - 2, 1)  # minus the first two bytes
+    # RLE and others with 1:1 fragment:frame
+    if len(fragment_offsets) == number_of_frames and not offset_values:
+        offset_values = fragment_offsets
+    if len(offset_values) != number_of_frames:
+        raise ValueError(
+            f"Number of frame items {len(offset_values)} does not match "
+            f"specified Number of Frames {number_of_frames}."
+        )
+    else:
+        basic_offset_table = offset_values
+
+    fp.seek(first_frame_location, 0)
+    return basic_offset_table
+
+
+def get_dataset_copy_with_frame_attrs(
+        original_dataset: Dataset,
+        optional_attributes: Tuple[str, ...] = _OPTIONAL_DATASET_ATTRIBUTES
+) -> Dataset:
+    """Create a copy of original_dataset with only the data elements needed
+    to parse frames
+
+    Parameters
+    ----------
+    original_dataset: Dataset
+        the dataset for which to create a minimal copy
+    optional_attributes: Tuple[str]
+
+    Returns
+    -------
+    Dataset
+        the copy of original_dataset
+
+    Raises
+    ------
+    ValueError
+        When original_dataset is missing attributes required for frame reading
+    """
+    ds_copy = Dataset()
+    ds_copy.file_meta = FileMetaDataset()
+    missing_attributes = list()
+    ds_copy.is_little_endian = original_dataset.is_little_endian
+    ds_copy.is_implicit_VR = original_dataset.is_implicit_VR
+    if not hasattr(original_dataset, "file_meta"):
+        transfer_syntax = None
+    else:
+        transfer_syntax = getattr(
+            original_dataset.file_meta, "TransferSyntaxUID", None
+        )
+    if transfer_syntax is None:
+        missing_attributes.append("file_meta.TransferSyntaxUID")
+    else:
+        ds_copy.file_meta.TransferSyntaxUID = transfer_syntax
+
+    for r_attribute in _REQUIRED_DATASET_ATTRIBUTES:
+        if not hasattr(original_dataset, r_attribute):
+            missing_attributes.append(r_attribute)
+        if missing_attributes:
+            continue
+        original_value = getattr(original_dataset, r_attribute)
+        setattr(ds_copy, r_attribute, original_value)
+
+    if missing_attributes:
+        del ds_copy
+        raise ValueError(
+            "Cannot make a copy of original_dataset as is missing the "
+            "following attributes required for frame reading: "
+            f"{missing_attributes}"
+        )
+    # not required, but need to copy if they exist
+    for attr in optional_attributes:
+        attr_value = original_dataset.get(attr)
+        setattr(ds_copy, attr, attr_value)
+    return ds_copy
+
+
+def decode_frame(
+        frame_bytes: bytes, original_dataset: Dataset
```

Comment:
I suggest passing the individual data elements to the function rather than `original_dataset`. That is more verbose, but easier to understand.

Specifically, I suggest using the existing implementation in highdicom: https://github.com/herrmannlab/highdicom/blob/d36e2a50350fd15531f549f52dafd87587a80afe/src/highdicom/frame.py#L327-L451

If that gets implemented here in pydicom, we can just import it in highdicom.

Reply:
Happy to refactor with the function from highdicom, but came across an issue - this [line](https://github.com/herrmannlab/highdicom/blob/d36e2a50350fd15531f549f52dafd87587a80afe/src/highdicom/frame.py#L449) has mismatched typing. I'm not terribly comfortable just telling mypy to ignore this.